src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -1062,8 +1066,8 @@ public static bool QueueUserWorkItem(WaitCallback callBack, object state)
             ExecutionContext context = ExecutionContext.Capture();
 
             object tpcallBack = (context == null || context.IsDefault) ?
-                new QueueUserWorkItemCallbackDefaultContext(callBack, state) :
-                (object)new QueueUserWorkItemCallback(callBack, state, context);
+                new QueueUserWorkItemCallbackDefaultContext(callBack!, state) :
+                (object)new QueueUserWorkItemCallback(callBack!, state, context);
```

Comment:
@jaredpar, I found the warnings here fairly strange.  `callBack` is defined to be non-nullable, but the compiler is warning me about potential null dereferences when passing it to another non-nullable.  I'm assuming that even though it's defined to be non-nullable, the compiler is seeing the null check earlier and assuming that means it could still be null, and then because we use ThrowHelper instead of a normal throw, it thinks a null could reach this location?   That's weird.  Is that by design or a bug?

Reply:
@stephentoub 

That behavior is by design. If you had an explicit `throw` here the code would work just fine. The use of the helper though means the compiler assumes execution continues. Hence after the conclusion of the `if` that `null` is a valid value for `callback`. 

This is probably the best example for why we need a "does not return" style annotation.

Reply:
So even though the symbol is non-nullable, because I checked it for null, the compiler then assumes it can be null? I get it, but it's counter intuitive to me.

Reply:
From the compiler point of view, you have execute ThrowArgumentNullException method if it is null and then kept going. The compiler does not do a global analysis required to figure out that ThrowArgumentNullException does not return.

If you change `ThrowHelper.ThrowArgumentNullException` to `throw new ArgumentNullException(...)`, this problem will disappear.

> This is probably the best example for why we need a "does not return" style annotation.

Or a better solution for ThrowHelpers...

Reply:
@jkotas, yes, I understand that.  I'm asking specifically about why the compiler thinks `callBack` may be null here, as it's defined to be non-nullable.  If I delete the `if (callBack == null)` check entirely, then it also doesn't warn here, so it appears to be inferring that `callBack` may be null, even though it was declared otherwise, because I checked against null.

Reply:
(It would be more intuitive to me if the compiler warned on the `callBack` in the `if (callBack == null)` comparison, saying that `callBack` was defined to be non-null and yet is being compared to null, such that I'd then write `if (callBack! == null)` instead.

Reply:
> such that I'd then write if (callBack! == null) instead.)

But then we would have to write all the null argument validation that throw (even if doesn't use throw helpers), to use the `!` forgiving operator, which doesn't seem ideal either.

Reply:
> But then we would have to write all the null argument validation that throw (even if doesn't use throw helpers), to use the ! forgiving operator, which doesn't seem ideal either.

Yup, I agree.  Just seems more "honest".